File: R/filter-metabolites.R
# Internal: two-step metabolite-level filter (CV first, then LOD rate).
# See DESIGN.md §4.3 and decisions.md #11-14.

# Coefficient of variation in percent. NA-safe; returns NA when the mean
# is zero or the input is degenerate.
.cv_pct <- function(x) {
  m <- mean(x, na.rm = TRUE)
  if (is.na(m) || m == 0) return(NA_real_)
  s <- stats::sd(x, na.rm = TRUE)
  100 * s / m
}

# CV filter on the QC matrix. Per-kit (default) requires both kits to
# pass; pooled treats all QC samples as one. Returns a logical pass
# vector (length = n_metabolite) and a long log frame with per-kit CVs.
.cv_filter <- function(qc_assay, qc_sample_meta,
                       cv_threshold = 20,
                       cv_mode = c("per_kit", "pooled")) {
  cv_mode <- match.arg(cv_mode)
  n_metab <- ncol(qc_assay)
  if (cv_mode == "pooled") {
    cv_values <- apply(qc_assay, 2L, .cv_pct)
    pass <- !is.na(cv_values) & cv_values <= cv_threshold
    log <- data.frame(
      cv_pooled = cv_values,
      cv_pass = pass,
      stringsAsFactors = FALSE
    )
    return(list(pass = pass, log = log))
  }

  if (!"kit_id" %in% names(qc_sample_meta)) {
    stop("CV per_kit mode requires `kit_id` in QC sample metadata.",
         call. = FALSE)
  }
  kit_ids <- sort(unique(stats::na.omit(qc_sample_meta$kit_id)))
  if (length(kit_ids) < 1L) {
    stop("No kits found in QC sample metadata.", call. = FALSE)
  }
  cv_by_kit <- vapply(kit_ids, function(k) {
    idx <- !is.na(qc_sample_meta$kit_id) & qc_sample_meta$kit_id == k
    if (!any(idx)) {
      return(rep(NA_real_, n_metab))
    }
    apply(qc_assay[idx, , drop = FALSE], 2L, .cv_pct)
  }, numeric(n_metab))
  if (length(kit_ids) == 1L) cv_by_kit <- matrix(cv_by_kit, ncol = 1L)
  pass <- apply(cv_by_kit, 1L, function(row) {
    all(!is.na(row) & row <= cv_threshold)
  })
  log_list <- stats::setNames(
    lapply(seq_along(kit_ids), function(k) cv_by_kit[, k]),
    paste0("cv_", kit_ids)
  )
  log_list$cv_pass <- pass
  log <- as.data.frame(log_list, stringsAsFactors = FALSE,
                       check.names = FALSE)
  list(pass = pass, log = log)
}

# Per-metabolite <LOD rate computed on the supplied (already-filtered)
# sample-status matrix. Pass when rate <= threshold.
.lod_rate_filter <- function(sample_status,
                             lod_rate_threshold = 0.5) {
  n <- nrow(sample_status)
  if (n == 0L) {
    stop("Cannot compute LOD rate from a zero-row status matrix.",
         call. = FALSE)
  }
  lod_rate <- colSums(sample_status == "< LOD", na.rm = TRUE) / n
  pass <- !is.na(lod_rate) & lod_rate <= lod_rate_threshold
  list(pass = pass, lod_rate = lod_rate)
}

File: python/metabosetr/filter_metabolites.py
"""Two-step metabolite filter (port of ``filter-metabolites.R``).

DESIGN.md §4.3. CV first (per-kit AND, or pooled), then LOD-rate. CV uses
the *sample* standard deviation (``ddof=1``) to match R ``sd()`` -- this is
the key numerical-fidelity point (numpy defaults to ``ddof=0``).
"""

from __future__ import annotations

import numpy as np
import pandas as pd


def _cv_pct(x: np.ndarray) -> float:
    """CV(%) = 100 * sd / mean; NA when mean is 0 or input is degenerate."""
    x = np.asarray(x, dtype=float)
    finite = x[~np.isnan(x)]
    if finite.size < 2:  # R sd() is NA for length < 2
        return np.nan
    m = finite.mean()
    if m == 0:
        return np.nan
    s = finite.std(ddof=1)  # sample sd, matches R sd()
    return 100 * s / m


def _cv_by_columns(assay: np.ndarray) -> np.ndarray:
    return np.array([_cv_pct(assay[:, j]) for j in range(assay.shape[1])])


def _cv_filter(
    qc_assay: np.ndarray,
    qc_sample_meta,
    cv_threshold: float = 20,
    cv_mode: str = "per_kit",
) -> dict:
    """CV filter on the QC matrix. Returns pass vector + long log frame."""
    if cv_mode not in ("per_kit", "pooled"):
        raise ValueError(f"Unknown cv_mode: {cv_mode}")
    n_metab = qc_assay.shape[1]

    if cv_mode == "pooled":
        cv_values = _cv_by_columns(qc_assay)
        cv_pass = (~np.isnan(cv_values)) & (cv_values <= cv_threshold)
        log = pd.DataFrame({"cv_pooled": cv_values, "cv_pass": cv_pass})
        return {"pass": np.asarray(cv_pass, dtype=bool), "log": log}

    if "kit_id" not in qc_sample_meta.columns:
        raise ValueError("CV per_kit mode requires `kit_id` in QC sample metadata.")
    kit_col = qc_sample_meta["kit_id"]
    kit_ids = sorted(k for k in kit_col.dropna().unique())
    if len(kit_ids) < 1:
        raise ValueError("No kits found in QC sample metadata.")

    cv_by_kit = np.column_stack(
        [
            _cv_by_columns(qc_assay[(kit_col == k).to_numpy(), :])
            if (kit_col == k).any()
            else np.full(n_metab, np.nan)
            for k in kit_ids
        ]
    )
    cv_pass = np.array(
        [
            bool(np.all((~np.isnan(row)) & (row <= cv_threshold)))
            for row in cv_by_kit
        ]
    )
    log_data = {f"cv_{k}": cv_by_kit[:, i] for i, k in enumerate(kit_ids)}
    log_data["cv_pass"] = cv_pass
    log = pd.DataFrame(log_data)
    return {"pass": np.asarray(cv_pass, dtype=bool), "log": log}


def _lod_rate_filter(sample_status: np.ndarray, lod_rate_threshold: float = 0.5) -> dict:
    """Per-metabolite ``<LOD`` rate on the (already-filtered) status matrix."""
    status = np.asarray(sample_status, dtype=object)
    n = status.shape[0]
    if n == 0:
        raise ValueError("Cannot compute LOD rate from a zero-row status matrix.")
    lod_rate = (status == "< LOD").sum(axis=0) / n
    lod_pass = (~np.isnan(lod_rate)) & (lod_rate <= lod_rate_threshold)
    return {"pass": np.asarray(lod_pass, dtype=bool), "lod_rate": lod_rate}
